File: backend/chikapi.ts
type HashSet<T extends number | string> = Record<T, undefined>
type HashMap<T extends number | string, U> = Record<T, U>
type Vec<T> = Array<T>
type Option<T> = T | undefined
type Result<T, U> = T | U
export interface Command {}
export type Method =
  | { t: "Get"; c: undefined }
  | { t: "Post"; c: undefined }
  | { t: "Put"; c: undefined }
  | { t: "Delete"; c: undefined }
export interface KeyField {
  prefix: Option<string>
  key: string
  size: Option<number>
}
export interface PathParameter {
  value: string
}
export interface Parameter {
  key: string
  value: string
}
export type QueryParameter =
  | { t: "Body"; c: Parameter }
  | { t: "Path"; c: PathParameter }
  | { t: "Query"; c: Parameter }
export type ReferenceParameter = { t: "Body"; c: string } | { t: "Path"; c: string } | { t: "Query"; c: string }
export type ReferenceParameters = Vec<ReferenceParameter>
export interface ReadModelRef {
  to: string
  parameters: ReferenceParameters
}
export type ReadModelStyle = { t: "LIST"; c: undefined } | { t: "TABLE"; c: undefined }
export interface ReadModel {
  path: string
  name: string
  id: string
  key_fields: Vec<KeyField>
  include: Option<Vec<string>>
  parameters: Vec<QueryParameter>
  reference: Option<ReadModelRef>
  root_key: Option<string>
  style: Option<ReadModelStyle>
}
export interface Schema {
  url: string
  has_openapi: boolean
  headers: Option<HashMap<string, string>>
  read_models: Vec<ReadModel>
  commands: Vec<Command>
  translation: HashMap<string, string>
}

File: src/schema/chikapi.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use serde_with::skip_serializing_none;

#[derive(Serialize, Deserialize, Clone, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct Command;
#[derive(Serialize, Deserialize, Clone, PartialEq)]
#[serde(rename_all = "camelCase")]
#[serde(untagged)]
pub enum Method {
    Get,
    Post,
    Put,
    Delete,
}

#[skip_serializing_none]
#[derive(Serialize, Deserialize, Clone, PartialEq, Eq, Hash)]
pub struct KeyField {
    pub prefix: Option<String>,
    pub key: String,
    pub size: Option<u8>, // 1-7
}

#[derive(Serialize, Deserialize, Clone, PartialEq)]
pub struct PathParameter {
    pub value: String,
}

#[derive(Serialize, Deserialize, Clone, PartialEq)]
pub struct Parameter {
    pub key: String,
    pub value: String,
}
#[derive(Serialize, Deserialize, Clone, PartialEq)]
#[serde(tag = "type")]
pub enum QueryParameter {
    Body(Parameter),
    Path(PathParameter),
    Query(Parameter),
}

#[derive(Serialize, Deserialize, Clone, PartialEq)]
#[serde(rename_all = "camelCase")]
#[serde(tag = "type", content = "key")]
pub enum ReferenceParameter {
    Body(String),
    Path(String),
    Query(String),
}

pub type ReferenceParameters = Vec<ReferenceParameter>;

#[derive(Serialize, Deserialize, Clone, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ReadModelRef {
    pub to: String,
    pub parameters: ReferenceParameters,
}

#[derive(Serialize, Deserialize, Clone, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum ReadModelStyle {
    LIST,
    TABLE,
}

#[skip_serializing_none]
#[derive(Serialize, Deserialize, Clone, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ReadModel {
    pub path: String,
    pub name: String,
    pub id: String,
    pub key_fields: Vec<KeyField>, // must be maximum value
    pub include: Option<Vec<String>>,
    pub parameters: Vec<QueryParameter>,
    pub reference: Option<ReadModelRef>,
    pub root_key: Option<String>,
    pub style: Option<ReadModelStyle>,
}

#[skip_serializing_none]
#[derive(Serialize, Deserialize, Clone, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct Schema {
    pub url: String,
    pub has_openapi: bool,
    pub headers: Option<HashMap<String, String>>,
    pub read_models: Vec<ReadModel>,
    pub commands: Vec<Command>,
    pub translation: HashMap<String, String>,
}

impl Parameter {
    pub fn new(k: String, v: String) -> Parameter {
        return Parameter { key: k, value: v };
    }
}

impl PathParameter {
    pub fn new(v: String) -> PathParameter {
        return PathParameter { value: v };
    }
}
